File: vulnlens_mobile/lib/services/calibration.dart
import '../models/models.dart';
import 'scorer.dart';

double? computeSpearmanCorrelation(List<double> ranksX, List<double> ranksY) {
  final n = ranksX.length;
  if (n <= 1 || ranksY.length != n) return null;

  var dSqSum = 0.0;
  for (var i = 0; i < n; i++) {
    final diff = ranksX[i] - ranksY[i];
    dSqSum += diff * diff;
  }

  final rho = 1.0 - (6.0 * dSqSum) / (n * (n * n - 1));
  return double.parse(rho.toStringAsFixed(4));
}

CalibrationReport evaluateGoldSet(
  List<CalibrationRecord> goldRecords,
  OrganizationProfile profile, {
  String practitionerField = 'practitioner_rank_bank',
}) {
  final scoredCandidates =
      <(CalibrationRecord, Vulnerability, ScoreBreakdown, int?)>[];

  for (final rec in goldRecords) {
    final vuln = Vulnerability(
      cveId: rec.cveId,
      productName: rec.productName,
      cvssBaseScore: rec.cvssBaseScore,
      cisaKev: rec.cisaKev,
      firstEpss: rec.firstEpss,
    );
    final breakdown = calculateScore(vuln, profile);
    final pRank = (practitionerField == 'practitioner_rank_startup')
        ? rec.practitionerRankStartup
        : rec.practitionerRankBank;
    scoredCandidates.add((rec, vuln, breakdown, pRank));
  }

  scoredCandidates.sort((a, b) {
    final scoreComp = b.$3.finalScore.compareTo(a.$3.finalScore);
    if (scoreComp != 0) return scoreComp;

    final cvssComp = (b.$2.cvssBaseScore ?? 0.0)
        .compareTo(a.$2.cvssBaseScore ?? 0.0);
    if (cvssComp != 0) return cvssComp;

    final epssComp =
        (b.$2.firstEpss ?? 0.0).compareTo(a.$2.firstEpss ?? 0.0);
    if (epssComp != 0) return epssComp;

    return a.$2.cveId.compareTo(b.$2.cveId);
  });

  final items = <GoldSetEvaluationItem>[];
  final engineRanks = <double>[];
  final practitionerRanks = <double>[];
  final absErrors = <double>[];

  for (var i = 0; i < scoredCandidates.length; i++) {
    final rankIdx = i + 1;
    final rec = scoredCandidates[i].$1;
    final vuln = scoredCandidates[i].$2;
    final breakdown = scoredCandidates[i].$3;
    final pRank = scoredCandidates[i].$4;
    final delta = pRank != null ? (rankIdx - pRank) : null;

    final notes = <String>[];
    if (breakdown.isCriticalProduct) notes.add('Critical Asset (1.4x)');
    if (vuln.cisaKev) notes.add('KEV Active');
    if ((vuln.firstEpss ?? 0.0) >= 0.5) {
      final epssPct = ((vuln.firstEpss ?? 0.0) * 100).toStringAsFixed(1);
      notes.add('High EPSS ($epssPct%)');
    }
    if ((vuln.cvssBaseScore ?? 0.0) >= 9.0) {
      notes.add('CVSS ${vuln.cvssBaseScore}');
    }

    final itemNotes =
        notes.isNotEmpty ? notes.join(', ') : 'Standard priority signals';

    items.add(
      GoldSetEvaluationItem(
        cveId: rec.cveId,
        productName: rec.productName,
        cvssBaseScore: rec.cvssBaseScore,
        cisaKev: rec.cisaKev,
        firstEpss: rec.firstEpss,
        scoreBreakdown: breakdown,
        engineRank: rankIdx,
        practitionerRank: pRank,
        rankDelta: delta,
        notes: itemNotes,
      ),
    );

    if (pRank != null) {
      engineRanks.add(rankIdx.toDouble());
      practitionerRanks.add(pRank.toDouble());
      absErrors.add((rankIdx - pRank).abs().toDouble());
    }
  }

  final corr = engineRanks.length >= 3
      ? computeSpearmanCorrelation(engineRanks, practitionerRanks)
      : null;
  final meanErr = absErrors.isNotEmpty
      ? (absErrors.reduce((a, b) => a + b) / absErrors.length)
      : null;

  String alignment;
  if (corr != null && corr >= 0.8) {
    alignment =
        'Strong rank alignment (Spearman ρ = ${corr.toStringAsFixed(2)}, Mean Rank Delta = ${meanErr?.toStringAsFixed(2)})';
  } else if (corr != null && corr >= 0.5) {
    alignment =
        'Moderate rank alignment (Spearman ρ = ${corr.toStringAsFixed(2)}, Mean Rank Delta = ${meanErr?.toStringAsFixed(2)})';
  } else {
    final corrStr = corr?.toStringAsFixed(2) ?? 'N/A';
    alignment = 'Calibration benchmark computed (Spearman ρ = $corrStr)';
  }

  final summary =
      'Gold-set sanity check against practitioner ranking for ${profile.name}. $alignment. Evaluated across ${items.length} curated ground-truth records.';

  return CalibrationReport(
    orgName: profile.name,
    orgId: profile.orgId,
    items: items,
    spearmanCorrelation: corr,
    meanAbsoluteRankError: meanErr,
    summaryText: summary,
  );
}

File: src/calibration.py
"""Gold set calibration and sanity check benchmarking engine."""

import math
from dataclasses import dataclass
from typing import List, Optional

from src.models import (
    CalibrationRecord,
    OrganizationProfile,
    ScoreBreakdown,
    Vulnerability,
)
from src.scorer import calculate_score


@dataclass
class GoldSetEvaluationItem:
    cve_id: str
    product_name: str
    cvss_base_score: float
    cisa_kev: bool
    first_epss: float
    score_breakdown: ScoreBreakdown
    engine_rank: int
    practitioner_rank: Optional[int]
    rank_delta: Optional[int]
    notes: str


@dataclass
class CalibrationReport:
    org_name: str
    org_id: str
    items: List[GoldSetEvaluationItem]
    spearman_correlation: Optional[float]
    mean_absolute_rank_error: Optional[float]
    summary_text: str


def compute_spearman_correlation(ranks_x: List[float], ranks_y: List[float]) -> Optional[float]:
    """Calculate Spearman's rank correlation coefficient between two equal-length rank arrays."""
    n = len(ranks_x)
    if n <= 1 or len(ranks_y) != n:
        return None
    d_sq_sum = sum((rx - ry) ** 2 for rx, ry in zip(ranks_x, ranks_y))
    rho = 1.0 - (6.0 * d_sq_sum) / (n * (n**2 - 1))
    return round(rho, 4)


def evaluate_gold_set(
    gold_records: List[CalibrationRecord],
    profile: OrganizationProfile,
    practitioner_field: str = "practitioner_rank_bank",
) -> CalibrationReport:
    """Evaluate the deterministic engine against a practitioner-ranked gold set baseline.
    
    CRITICAL CONSTRAINT: This function evaluates ONLY the isolated gold set records.
    It never mixes or pollutes the production vulnerability dataset.
    """
    # 1. Convert gold records to temporary Vulnerability objects
    scored_candidates = []
    for rec in gold_records:
        vuln = Vulnerability(
            cve_id=rec.cve_id,
            product_name=rec.product_name,
            cvss_base_score=rec.cvss_base_score,
            cisa_kev=rec.cisa_kev,
            first_epss=rec.first_epss,
        )
        breakdown = calculate_score(vuln, profile)
        
        # Get target practitioner rank based on field
        p_rank = getattr(rec, practitioner_field, None)
        scored_candidates.append((rec, vuln, breakdown, p_rank))

    # 2. Sort by engine final_score descending
    scored_candidates.sort(
        key=lambda x: (
            -x[2].final_score,
            -(x[1].cvss_base_score or 0.0),
            -(x[1].first_epss or 0.0),
            x[1].cve_id,
        )
    )

    # 3. Build evaluation items
    items: List[GoldSetEvaluationItem] = []
    engine_ranks: List[float] = []
    practitioner_ranks: List[float] = []
    abs_errors: List[float] = []

    for rank_idx, (rec, vuln, breakdown, p_rank) in enumerate(scored_candidates, start=1):
        delta = (rank_idx - p_rank) if p_rank is not None else None
        
        # Build diagnostic notes
        notes = []
        if breakdown.is_critical_product:
            notes.append("Critical Asset (1.4x)")
        if vuln.cisa_kev:
            notes.append("KEV Active")
        if (vuln.first_epss or 0.0) >= 0.5:
            notes.append(f"High EPSS ({vuln.first_epss:.1%})")
        if (vuln.cvss_base_score or 0.0) >= 9.0:
            notes.append(f"CVSS {vuln.cvss_base_score}")

        item_note = ", ".join(notes) if notes else "Standard priority signals"

        items.append(
            GoldSetEvaluationItem(
                cve_id=rec.cve_id,
                product_name=rec.product_name,
                cvss_base_score=rec.cvss_base_score,
                cisa_kev=rec.cisa_kev,
                first_epss=rec.first_epss,
                score_breakdown=breakdown,
                engine_rank=rank_idx,
                practitioner_rank=p_rank,
                rank_delta=delta,
                notes=item_note,
            )
        )

        if p_rank is not None:
            engine_ranks.append(float(rank_idx))
            practitioner_ranks.append(float(p_rank))
            abs_errors.append(abs(rank_idx - p_rank))

    # 4. Compute metrics
    corr = compute_spearman_correlation(engine_ranks, practitioner_ranks) if len(engine_ranks) >= 3 else None
    mean_err = (sum(abs_errors) / len(abs_errors)) if abs_errors else None

    if corr is not None and corr >= 0.8:
        alignment = f"Strong rank alignment (Spearman ρ = {corr:.2f}, Mean Rank Delta = {mean_err:.2f})"
    elif corr is not None and corr >= 0.5:
        alignment = f"Moderate rank alignment (Spearman ρ = {corr:.2f}, Mean Rank Delta = {mean_err:.2f})"
    else:
        corr_str = f"{corr:.2f}" if corr is not None else "N/A"
        alignment = f"Calibration benchmark computed (Spearman ρ = {corr_str})"

    summary = (
        f"Gold-set sanity check against practitioner ranking for {profile.name}. "
        f"{alignment}. Evaluated across {len(items)} curated ground-truth records."
    )

    return CalibrationReport(
        org_name=profile.name,
        org_id=profile.org_id,
        items=items,
        spearman_correlation=corr,
        mean_absolute_rank_error=mean_err,
        summary_text=summary,
    )
